pkg/providers/ovirt/client/ovirt-client.go

```diff
@@ -0,0 +1,270 @@
+package ovirtclient
+
+import (
+	"fmt"
+
+	ovirtsdk "github.com/ovirt/go-ovirt"
+)
+
+// OvirtClient retrieves rich VM object - with all necessary links have been followed
+type OvirtClient interface {
+	GetVM(connectionSettings *ConnectionSettings, id *string, name *string, cluster *string) (*ovirtsdk.Vm, error)
+}
+
+// ConnectionSettings wrap information required to make oVirt API connection
+type ConnectionSettings struct {
+	URL      string
+	Username string
+	Password string
+	Insecure bool
+}
+
+// RichOvirtClient is responsilbe for retrieving VM data from oVirt API
+type RichOvirtClient struct {
+}
+
+// GetVM rertrieves oVirt VM data for given id or name and cluster. VM will have certain links followed and updated.
+func (client *RichOvirtClient) GetVM(cs *ConnectionSettings, id *string, name *string, cluster *string) (*ovirtsdk.Vm, error) {
+	con, err := client.createConnection(cs.URL, cs.Username, cs.Password, cs.Insecure)
+	if err != nil {
+		return nil, err
+	}
+	vm, err := client.fetchVM(con, id, name, cluster)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateNics(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateDiskAttachments(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateCdRoms(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateFloppies(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateHostDevices(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateReportedDevices(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateQuota(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateWatchdogs(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateGrapthicsConsoles(con, vm)
+	if err != nil {
+		return nil, err
+	}
+	return vm, nil
+}
+
+func (client *RichOvirtClient) fetchVM(con *ovirtsdk.Connection, id *string, name *string, cluster *string) (*ovirtsdk.Vm, error) {
+	// Id of the VM specified:
+	if id != nil {
+		response, err := con.SystemService().VmsService().VmService(*id).Get().Send()
+		if err != nil {
+			return nil, err
+		}
+		vm, _ := response.Vm()
+		return vm, nil
+	}
+	// Cluster/name of the VM specified:
+	response, err := con.SystemService().VmsService().List().Search(fmt.Sprintf("name=%v and cluster=%v", *name, *cluster)).Send()
+	if err != nil {
+		return nil, err
+	}
+	vms, _ := response.Vms()
+	if len(vms.Slice()) > 0 {
+		return vms.Slice()[0], nil
+	}
+	return nil, fmt.Errorf("Virtual machine %v not found", *name)
+}
+
+func (client *RichOvirtClient) populateHostDevices(con *ovirtsdk.Connection, vm *ovirtsdk.Vm) error {
+	if hdcs, ok := vm.HostDevices(); ok {
+		if _, ok := hdcs.Href(); ok {
```

Comment:
What is the value of `Href` check here? It is done as second check in `FollowLink` function.

Reply:
`FollowLink` returns error due to several reasons and that's the only one I want to ignore without relying on error messages.

Reply:
You still return error if something is not right. Not sure why to handle something twice :thinking:

Reply:
No, if there is no href, I don't follow the link. If I followed, the sdk method would return an error, which I would have to dispatch based on the message - to ignore no-href error and propagate others.

Reply:
OK, it is up to you. You could save good number of lines if you rely on `FollowLink` to check it.

Reply:
Actually, I would have to decide what to do with errors for no href, as that is not an import-breaker.
The core reason for that href pre-check is that I've gotten an error while trying to retrieve quota for a VM from a real, live cluster and that error should not be a blocker.

Reply:
Most of the object must have the href. What was the specific case that your quota didn't have href? If it really happened it's most probably oVirt bug.

Reply:
Actually, the documentation say it is a link, but in the VM xml I just inspected it is just an element:
`  <quota id="3d1bb0a1-785c-4fbe-8b0b-fdb8bc4ae5d7"/>`.
Docs: http://ovirt.github.io/ovirt-engine-api-model/4.4/#types/vm.

Reply:
Seeing that I am not 100% confident that removing all the other checks will always be safe.

Reply:
I think we should trust the API, this is indeed bug, and I've opened the bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1814613. I think we should remove the href checks and add it only for quota with comment explaining why it's there, and link to bug. But feel free to leave as is.

Reply:
Thanks! I will remove all the unnecessary checks.